tools/Az.Tools.Predictor/Az.Tools.Predictor/AzContext.cs

```diff
@@ -0,0 +1,177 @@
+// ----------------------------------------------------------------------------------
+//
+// Copyright Microsoft Corporation
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using System.Management.Automation;
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Net.NetworkInformation;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Microsoft.Azure.PowerShell.Tools.AzPredictor
+{
+    using PowerShell = System.Management.Automation.PowerShell;
+
+    /// <summary>
+    /// The class for the current Azure PowerShell context.
+    /// </summary>
+    internal sealed class AzContext : IAzContext
+    {
+        /// <inheritdoc/>
+        public Version AzVersion { get; private set; } = new Version("0.0.0.0");
+
+        /// <inheritdoc/>
+        public string HashedUserId { get; private set; } = string.Empty;
+
+        private string _cachedHashedMacAddress;
+        /// <inheritdoc/>
+        public string HashedMacAddress
+        {
+            get
+            {
+                if (_cachedHashedMacAddress == null)
+                {
+                    _cachedHashedMacAddress = string.Empty;
+
+                    var macAddress = GetMACAddress();
+                    if (!string.IsNullOrWhiteSpace(macAddress))
+                    {
+                        _cachedHashedMacAddress = GenerateSha256HashString(macAddress)?.Replace("-", string.Empty).ToLowerInvariant();
+                    }
+                }
+
+                return _cachedHashedMacAddress;
+            }
+        }
+
+        /// <inheritdoc/>
+        public void UpdateContext()
+        {
+            AzVersion = GetAzVersion();
+            HashedUserId = GenerateSha256HashString(GetUserAccountId());
+        }
+
+        /// <summary>
+        /// Gets the latest version from the loaded Az modules.
+        /// </summary>
+        private Version GetAzVersion()
```

Comment:
I don't suggest to get Az version again because the logic cannot cover all cases. Version cannot be fetched on Windows PowerShell. Or user may use Azure PowerShell module but without Az.

@dcaro , what's your thought?

Reply:
I think that the version of the module (Az.xxxx and not Az) is more relevant.

Reply:
@dcaro , please check if current implementation is the same as your suggestion.

It is trying to record all az.xxx versions.